File: 05_Word_Cruncher.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
#include <algorithm>

using namespace std;

vector<string> results;

void dfs(const vector<string>& words, const string& target, vector<int>& path, int start, unordered_set<int>& used) {
    if (start == target.size()) {
        string result;
        for (int i : path) {
            result += words[i] + " ";
        }
        if (std::find(results.begin(), results.end(), result) == results.end()) {
        results.push_back(result);
        }

        return;
    }
    for (int i = 0; i < words.size(); i++) {
        if (used.find(i) == used.end() && target.substr(start, words[i].size()) == words[i]) {
            used.insert(i);
            path.push_back(i);
            dfs(words, target, path, start + words[i].size(), used);
            path.pop_back();
            used.erase(i);
        }
    }
}

vector<string> word_cruncher(const vector<string>& words, const string& target) {
    results.clear();
    vector<int> path;
    unordered_set<int> used;
    dfs(words, target, path, 0, used);
    return results;
}

int main() {
    vector<string> words;
    string target;

    string input;
    getline(cin, input);
    size_t pos = 0;
    while ((pos = input.find(", ")) != string::npos) {
        words.push_back(input.substr(0, pos));
        input.erase(0, pos + 2);
    }
    words.push_back(input);

    getline(cin, target);

    vector<string> result = word_cruncher(words, target);
    for (const string& res : result) {
        cout << res << endl;
    }

    return 0;
}

File: 05_Word_Cruncher.py
def word_cruncher(words, target):
    results = []

    def dfs(path, start, used):
        if start == len(target):
            result = ' '.join(words[i] for i in path)
            if result not in results:
                results.append(result)
            return
        for i in range(len(words)):
            if i not in used and target[start:start + len(words[i])] == words[i]:
                used.add(i)
                dfs(path + [i], start + len(words[i]), used)
                used.remove(i)

    dfs([], 0, set())

    return results

words = input().split(', ')
target = input()
print('\n'.join(word_cruncher(words, target)))
